qiita_db/artifact.py

```diff
@@ -1115,23 +1119,98 @@ def descendants_with_jobs(self):
             sql = """SELECT *
                      FROM qiita.artifact_descendants_with_jobs(%s)"""
             qdb.sql_connection.TRN.add(sql, [self.id])
-            edges = qdb.sql_connection.TRN.execute_fetchindex()
+            sql_edges = qdb.sql_connection.TRN.execute_fetchindex()
 
-        lineage = nx.DiGraph()
-        if edges:
+            lineage = nx.DiGraph()
+            edges = set()
             nodes = {}
-            for jid, pid, cid in edges:
-                if jid not in nodes:
-                    nodes[jid] = ('job', qdb.processing_job.ProcessingJob(jid))
-                if pid not in nodes:
-                    nodes[pid] = ('artifact', qdb.artifact.Artifact(pid))
-                if cid not in nodes:
-                    nodes[cid] = ('artifact', qdb.artifact.Artifact(cid))
-
-                lineage.add_edge(nodes[pid], nodes[jid])
-                lineage.add_edge(nodes[jid], nodes[cid])
-        else:
-            lineage.add_node(('artifact', self))
+            if sql_edges:
+                for jid, pid, cid in sql_edges:
+                    if jid not in nodes:
+                        nodes[jid] = ('job',
+                                      qdb.processing_job.ProcessingJob(jid))
+                    if pid not in nodes:
+                        nodes[pid] = ('artifact', qdb.artifact.Artifact(pid))
+                    if cid not in nodes:
+                        nodes[cid] = ('artifact', qdb.artifact.Artifact(cid))
+                    edges.add((nodes[pid], nodes[jid]))
+                    edges.add((nodes[jid], nodes[cid]))
+            else:
+                nodes[self.id] = ('artifact', self)
+                lineage.add_node(nodes[self.id])
+
+            # The code above returns all the jobs that have been successfully
+            # executed. We need to add all the jobs that are in all the other
+            # status. Approach: Loop over all the artifacts and add all the
+            # jobs that have been attached to them.
+            visited = set()
+            queue = nodes.keys()
+            while queue:
+                current = queue.pop(0)
+                if current not in visited:
+                    visited.add(current)
```

Comment:
should this be the last line within this if? Just worry that something fails in the if/elifs and we still add the current.

Reply:
I've added an else that raises an error if `n_type`is not in `['artifact', 'job', 'type']`. That should never happen but better be safe than sorry.